rust/kernel/static_assert.rs

```diff
@@ -4,35 +4,198 @@
+/// static_assert!((SOME_CONSTANT_DEPENDING_ON_ARCH) is in {FOO, BAR, BAZ});
+/// static_assert!((core::mem::size_of::<usize>()) is in {4, 8});
+/// ```
+///
+/// # Interval membership assertion: `(expr) is in [min, max]`
+///
+/// Statically asserts that the given expression (typically a `const` integer) is in a closed
+/// interval (i.e. inclusive range). The negated form (`is not in`) is also available.
+///
+/// ## Examples
+///
+/// ```
+/// // Trivial usage.
+/// static_assert!((-2) is not in [-1, 2]);
+/// static_assert!((-1) is     in [-1, 2]);
+/// static_assert!(( 0) is     in [-1, 2]);
+/// static_assert!(( 1) is     in [-1, 2]);
+/// static_assert!(( 2) is     in [-1, 2]);
+/// static_assert!(( 3) is not in [-1, 2]);
+///
+/// // Typical usage.
+/// static_assert!((FOO) is in [MIN_FOO, MAX_FOO]);
+/// ```
+///
+/// # Fits-in-type assertion: `(expr) fits in type`
+///
+/// Statically asserts that the given expression (typically a `const` integer) fits in the given
+/// type (which must provide `T::MIN` and `T::MAX`). The negated form (`does not fit in`) is also
+/// available.
+///
+/// Casting a "kernel integer" (i.e. up to [`i64`]/[`u64`]) to [`i128`] within the expression is
+/// allowed to easily manipulate integers: no 128-bit code will be generated since it will be
+/// evaluated in a const context.
+///
+/// ## Examples
+///
+/// ```
+/// // Trivial usage.
+/// static_assert!(( -1) does not fit  in u8);
+/// static_assert!((  0)          fits in u8);
+/// static_assert!((255)          fits in u8);
+/// static_assert!((256) does not fit  in u8);
+///
+/// // Two's complement.
+/// static_assert!((-128)          fits in i8);
+/// static_assert!(( 127)          fits in i8);
+/// static_assert!(( 128) does not fit  in i8);
+///
+/// // Using `i128` for easy manipulation of integers.
+/// const MAX_ERRNO: u32 = 4095;
+/// static_assert!((-(MAX_ERRNO as i128)) fits in i16);
+/// ```
 #[macro_export]
 macro_rules! static_assert {
+    // Boolean assertion: `expr`.
     ($condition:expr) => {
         // Based on the latest one in `rustc`'s one before it was [removed].
         //
         // [removed]: https://github.com/rust-lang/rust/commit/c2dad1c6b9f9636198d7c561b47a2974f5103f6d
         #[allow(dead_code)]
         const _: () = [()][!($condition) as usize];
     };
+
+    // Set membership assertion: `(expr) is in {a0, ..., aN}`.
+    (($expression:expr) is in {$($a:expr),+}) => {
+        static_assert!(  $(($expression) == ($a))||* );
+    };
+    (($expression:expr) is not in {$($a:expr),+}) => {
+        static_assert!(!($(($expression) == ($a))||*));
+    };
+
+    // Interval membership assertion: `(expr) is in [min, max]`.
+    (($expression:expr) is in [$min:expr, $max:expr]) => {
+        static_assert!(  ($expression) >= ($min) && ($expression) <= ($max) );
+    };
+    (($expression:expr) is not in [$min:expr, $max:expr]) => {
+        static_assert!(!(($expression) >= ($min) && ($expression) <= ($max)));
+    };
```

Comment:
It would probably be better to use regular Rust range syntax (and [range].contains(...)) here. I think it's generally useful to have macros resemble normal syntax as closely as possible as it makes the macros easier to learn. Is there a reason we can't / don't want to do that here?